logstash-core/lib/logstash/plugins/registry.rb

```diff
@@ -1,86 +1,243 @@
+        is_a_plugin?(plugin_klass, plugin_name)
+      end
+    rescue LoadError, NameError => e
+      logger.debug("Problems loading the plugin with", :type => type, :name => name)
+      raise(LogStash::PluginLoadingError, I18n.t("logstash.pipeline.plugin-loading-error", :type => type, :name => name, :error => e.to_s))
+    end
+
+    def lazy_add(type, name, klass)
+      logger.error("On demand adding plugin to the registry", :name => name, :type => type, :class => klass)
+      add_plugin(type, name, klass)
+    end
+
+    def add(type, name, klass)
+      logger.error("Adding plugin to the registry", :name => name, :type => type, :class => klass)
+      add_plugin(type, name, klass)
+    end
+
+    def get(type, plugin_name)
+      @registry[key_for(type, plugin_name)]
+    end
+
+    def exists?(type, name)
+      @registry.include?(key_for(type, name))
+    end
+
+    def size
+      @registry.size
+    end
+
+    private
+    # lookup a plugin by type and name in the existing LogStash module namespace
+    # ex.: namespace_lookup("filter", "grok") looks for LogStash::Filters::Grok
+    # @param type [String] plugin type, "input", "ouput", "filter"
+    # @param name [String] plugin name, ex.: "grok"
+    # @return [Class] the plugin class or raises NameError
+    # @raise NameError if plugin class does not exist or is invalid
+    def namespace_lookup(type, name)
+      type_const = "#{type.capitalize}s"
+      namespace = LogStash.const_get(type_const)
+      # the namespace can contain constants which are not for plugins classes (do not respond to :config_name)
+      # namespace.constants is the shallow collection of all constants symbols in namespace
+      # note that below namespace.const_get(c) should never result in a NameError since c is from the constants collection
+      klass_sym = namespace.constants.find { |c| is_a_plugin?(namespace.const_get(c), name) }
+      klass = klass_sym && namespace.const_get(klass_sym)
+
+      raise(NameError) unless klass
+      klass
+    end
+
+    # check if klass is a valid plugin for name
+    # @param klass [Class] plugin class
+    # @param name [String] plugin name
+    # @return [Boolean] true if klass is a valid plugin for name
+    def is_a_plugin?(klass, name)
+      klass.ancestors.include?(LogStash::Plugin) && klass.respond_to?(:config_name) && klass.config_name == name
+    end
+
+    def add_plugin(type, name, klass)
+      if !exists?(type, name)
+        specification_klass = type == :universal ? UniversalPluginSpecification : PluginSpecification
+        @registry[key_for(type, name)] = specification_klass.new(type, name, klass)
+      else
+        logger.info("Ignoring, plugin already added to the registry", :name => name, :type => type, :klass => klass)
+      end
     end
 
-    def registered?(path)
-      @registry.has_key?(path)
+    def key_for(type, plugin_name)
+      "#{type}-#{plugin_name}"
     end
+  end end
 
-  end
+  PluginRegistry = Plugins::Registry.new
```

Comment:
Well `PluginRegistry` is a constants, it just not Capitalized.


*I double checked that*